tests/lib/utils/test-alerts.php

```diff
@@ -0,0 +1,258 @@
+		$validate_message_method = self::get_alerts_method( 'validate_message' );
+
+		$result = $validate_message_method->invokeArgs( $alerts, [ $message ] );
+
+		$this->assertWPError( $result );
+		$this->assertEquals( 'invalid-alert-message', $result->get_error_code() );
+		$this->assertEquals( 'test.host', $alerts->service_address, 'Wrong alerts service address' );
+		$this->assertEquals( 9999, $alerts->service_port, 'Wrong alerts service port' );
+		$this->assertEquals( 'http://test.host:9999/v1.0/alerts', $alerts->service_url, 'Wrong alerts service URL' );
+	}
+
+	/**
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
+	 */
+	public function test__validate_details() {
+		define( 'ALERT_SERVICE_ADDRESS', 'test.host' );
+		define( 'ALERT_SERVICE_PORT', 9999 );
+
+		$details = [
+			'alias' => 'test/alert',
+			'description' => 'Test alert',
+			'entity' => 'test',
+			'priority' => 'P4',
+			'source' => 'test',
+		];
+
+		$alerts = Alerts::instance();
+
+		$validate_details_method = self::get_alerts_method( 'validate_details' );
+
+		$result = $validate_details_method->invokeArgs( $alerts, [ $details ] );
+
+		$this->assertEquals( $details, $result );
+		$this->assertEquals( 'test.host', $alerts->service_address, 'Wrong alerts service address' );
+		$this->assertEquals( 9999, $alerts->service_port, 'Wrong alerts service port' );
+		$this->assertEquals( 'http://test.host:9999/v1.0/alerts', $alerts->service_url, 'Wrong alerts service URL' );
+	}
+
+	public function get_test_data__invalid_details() {
+		return [
+			'invalid-type' => [ 'string' ],
+			'missing-keys' => [ 
+				'alias' => 'test/alert',
+				'description' => 'Test alert',
+				'entity' => 'test',
+				'source' => 'test',
+			],
+			'extra-keys' => [ 
+				'alias' => 'test/alert',
+				'description' => 'Test alert',
+				'entity' => 'test',
+				'priority' => 'P4',
+				'source' => 'test',
+				'extra' => 'invalid',
+			],
+			'empty-keys' => [
+				'alias' => 'test/alert',
+				'description' => '',
+				'entity' => 'test',
+				'priority' => 'P4',
+				'source' => 'test',
+			],
+		];
+	}
+
+	/**
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
+	 * @dataProvider get_test_data__invalid_details
+	 */
+	public function test__validate_details__invalid_details( $details ) {
+		define( 'ALERT_SERVICE_ADDRESS', 'test.host' );
+		define( 'ALERT_SERVICE_PORT', 9999 );
+
+		$alerts = Alerts::instance();
+
+		$validate_details_method = self::get_alerts_method( 'validate_details' );
+
+		$result = $validate_details_method->invokeArgs( $alerts, [ $details ] );
+
+		$this->assertWPError( $result );
+		$this->assertEquals( 'invalid-opsgenie-details', $result->get_error_code() );
+		$this->assertEquals( 'test.host', $alerts->service_address, 'Wrong alerts service address' );
+		$this->assertEquals( 9999, $alerts->service_port, 'Wrong alerts service port' );
+		$this->assertEquals( 'http://test.host:9999/v1.0/alerts', $alerts->service_url, 'Wrong alerts service URL' );
+	}
+}
```

Comment:
It would be worth adding a test case for the rate limiting too to ensure that's working as expected.

Reply:
I think this does not apply anymore after last refactoring.

Reply:
Well we still don't have any test that proves that if you try to send more than 1 message of `$kind` within the `$interval` time, it fails, so we don't _know_ that the rate limiting is working and won't regress in the future. But I don't want to hold this up any longer so let's ship what we have.